File: archives/APPLE/shared/UI/Rendering/MarkdownWebView.swift
import SwiftUI

/// Renders a markdown string with KaTeX math via a WKWebView hosting the
/// katex-shell.html bundled with the package.
///
/// Optional `tableName` + `highlightedRows` enable curriculum-specific
/// table rendering: the JS shell will prepend a full-width header row
/// with the table name, remove the original column-header row, merge
/// identical first-column cells into rowspans, and apply `.highlight` to
/// data rows by index.
///
/// The view sizes itself to the rendered content via a WKScriptMessage
/// channel (`contentHeight`) — katex-shell.html posts
/// `document.body.scrollHeight` after each render and SwiftUI frames the
/// outer view to that height. Internal WebKit scrolling is disabled so
/// multiple tables on one screen flow naturally inside an outer
/// ScrollView without competing scroll gestures.
///
/// On hosts without WebKit (watchOS, and `swift build` running on
/// macOS host-only for type-checking) this falls back to a plain
/// SwiftUI stub so call sites can embed it from shared code. The
/// real rendering path is iOS-only. The guard is `canImport(WebKit)`
/// rather than `canImport(UIKit)` because watchOS has UIKit but NOT
/// WebKit — the narrower check is what lets `ScienceCoreUI` compile
/// cleanly for the watchOS slice of the package.
#if canImport(WebKit)

import UIKit
import WebKit

public struct MarkdownWebView: View {
    let markdown: String
    let tableName: String?
    let highlightedRows: [Int]

    @State private var contentHeight: CGFloat = 80

    public init(markdown: String, tableName: String? = nil, highlightedRows: [Int] = []) {
        self.markdown = markdown
        self.tableName = tableName
        self.highlightedRows = highlightedRows
    }

    public var body: some View {
        MarkdownWebViewRep(
            markdown: markdown,
            tableName: tableName,
            highlightedRows: highlightedRows,
            contentHeight: $contentHeight
        )
        .frame(height: contentHeight)
    }
}

private struct MarkdownWebViewRep: UIViewRepresentable {
    let markdown: String
    let tableName: String?
    let highlightedRows: [Int]
    @Binding var contentHeight: CGFloat

    struct RenderOptions {
        let markdown: String
        let tableName: String?
        let highlightedRows: [Int]
    }

    func makeUIView(context: Context) -> WKWebView {
        let config = WKWebViewConfiguration()
        // JS → Swift bridge: katex-shell.html posts scrollHeight here
        // after each render so we can size the outer SwiftUI frame.
        config.userContentController.add(context.coordinator, name: "contentHeight")

        let view = WKWebView(frame: .zero, configuration: config)
        view.isOpaque = false
        view.backgroundColor = .clear
        view.scrollView.backgroundColor = .clear
        // Disable internal scrolling — the SwiftUI parent ScrollView owns
        // the scroll gesture, and the webview sizes itself to the
        // contentHeight binding so nothing is clipped.
        view.scrollView.isScrollEnabled = false
        view.scrollView.bounces = false

        if let url = Bundle.module.url(forResource: "katex-shell", withExtension: "html") {
            view.loadFileURL(url, allowingReadAccessTo: url.deletingLastPathComponent())
        }
        context.coordinator.webView = view
        context.coordinator.pendingOptions = RenderOptions(
            markdown: markdown,
            tableName: tableName,
            highlightedRows: highlightedRows
        )
        view.navigationDelegate = context.coordinator
        return view
    }

    func updateUIView(_ uiView: WKWebView, context: Context) {
        context.coordinator.render(
            RenderOptions(
                markdown: markdown,
                tableName: tableName,
                highlightedRows: highlightedRows
            )
        )
    }

    func makeCoordinator() -> Coordinator {
        Coordinator(contentHeight: $contentHeight)
    }

    final class Coordinator: NSObject, WKNavigationDelegate, WKScriptMessageHandler {
        weak var webView: WKWebView?
        var pendingOptions: RenderOptions?
        var loaded = false
        var contentHeight: Binding<CGFloat>

        init(contentHeight: Binding<CGFloat>) {
            self.contentHeight = contentHeight
        }

        func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
            loaded = true
            if let pending = pendingOptions {
                render(pending)
            }
        }

        /// Called by katex-shell.html via
        /// `window.webkit.messageHandlers.contentHeight.postMessage(...)`
        /// at the end of every render. Pushes the new height into the
        /// SwiftUI @State binding so the view frame updates.
        func userContentController(
            _ userContentController: WKUserContentController,
            didReceive message: WKScriptMessage
        ) {
            guard message.name == "contentHeight" else { return }
            if let number = message.body as? NSNumber {
                let height = CGFloat(truncating: number)
                if height > 0 {
                    DispatchQueue.main.async {
                        self.contentHeight.wrappedValue = height
                    }
                }
            }
        }

        func render(_ options: RenderOptions) {
            guard let view = webView else { return }
            guard loaded else { pendingOptions = options; return }

            var payload: [String: Any] = [
                "markdown": options.markdown,
                "highlightedRows": options.highlightedRows
            ]
            if let tableName = options.tableName {
                payload["tableName"] = tableName
            }

            guard let data = try? JSONSerialization.data(withJSONObject: payload, options: [.fragmentsAllowed]),
                  let json = String(data: data, encoding: .utf8) else {
                return
            }
            view.evaluateJavaScript("window.renderMarkdown(\(json));", completionHandler: nil)
        }
    }
}

#else

/// Host-only stub so `swift build` on macOS type-checks. Never runs in
/// practice — the app is iOS-only.
public struct MarkdownWebView: View {
    let markdown: String
    let tableName: String?
    let highlightedRows: [Int]

    public init(markdown: String, tableName: String? = nil, highlightedRows: [Int] = []) {
        self.markdown = markdown
        self.tableName = tableName
        self.highlightedRows = highlightedRows
    }

    public var body: some View {
        Text("Preview unavailable on this platform")
            .foregroundStyle(.secondary)
            .padding()
    }
}

#endif

File: archives/android/ui/src/main/java/com/vivianweidai/science/ui/MarkdownWebView.kt
package com.vivianweidai.science.ui

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import org.json.JSONArray
import org.json.JSONObject

/**
 * Renders a markdown string with KaTeX math via a WebView hosting the
 * bundled `katex-shell.html` asset.
 *
 * Optional [tableName] + [highlightedRows] enable curriculum-specific
 * table rendering: the JS shell prepends a full-width header row with
 * the table name, removes the original column-header row, merges
 * identical first-column cells into rowspans, and applies `.highlight`
 * to data rows by index.
 *
 * The view sizes itself to rendered content via a JavaScriptInterface
 * channel (`contentHeight`) — the shell posts `document.body.scrollHeight`
 * after each render and Compose frames the outer view to that height.
 * Internal WebView scrolling is disabled so multiple tables on one
 * screen flow naturally inside an outer LazyColumn.
 */
@SuppressLint("SetJavaScriptEnabled")
@Composable
fun MarkdownWebView(
    markdown: String,
    tableName: String? = null,
    highlightedRows: List<Int> = emptyList(),
    onImageTap: ((String) -> Unit)? = null,
) {
    val context = LocalContext.current
    var contentHeight: Dp by remember { mutableStateOf(80.dp) }

    AndroidView(
        modifier = Modifier.fillMaxWidth().height(contentHeight),
        factory = {
            WebView(context).apply {
                layoutParams = ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                )
                isVerticalScrollBarEnabled = false
                isHorizontalScrollBarEnabled = false
                overScrollMode = WebView.OVER_SCROLL_NEVER
                setBackgroundColor(0) // transparent
                settings.javaScriptEnabled = true
                settings.domStorageEnabled = true
                settings.allowFileAccess = true
                settings.loadWithOverviewMode = true

                addJavascriptInterface(
                    object {
                        @JavascriptInterface
                        fun postHeight(h: Double) {
                            // Called from JS on every render. Marshal back to
                            // the main thread for Compose state update.
                            post {
                                if (h > 0) contentHeight = h.dp
                            }
                        }
                    },
                    "AndroidBridge",
                )

                webViewClient = object : WebViewClient() {
                    private var loaded = false
                    private var pending: Triple<String, String?, List<Int>>? = null

                    override fun onPageFinished(view: WebView, url: String?) {
                        loaded = true
                        pending?.let { (md, tn, hr) ->
                            render(view, md, tn, hr)
                            pending = null
                        } ?: render(view, markdown, tableName, highlightedRows)
                    }

                    override fun shouldOverrideUrlLoading(
                        view: WebView,
                        req: WebResourceRequest,
                    ): Boolean {
                        val url = req.url.toString()
                        val lower = url.lowercase()
                        val isImage = lower.endsWith(".jpg") || lower.endsWith(".jpeg") ||
                                      lower.endsWith(".png") || lower.endsWith(".gif")
                        return if (isImage && onImageTap != null) {
                            onImageTap.invoke(url); true
                        } else {
                            context.startActivity(
                                Intent(Intent.ACTION_VIEW, Uri.parse(url))
                                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                            )
                            true
                        }
                    }
                }

                // Load the KaTeX shell from assets. The JS then receives
                // markdown payloads via `window.renderMarkdown(...)`.
                loadUrl("file:///android_asset/katex-shell.html")
            }
        },
        update = { view ->
            render(view, markdown, tableName, highlightedRows)
        },
    )
}

private fun render(view: WebView, markdown: String, tableName: String?, highlightedRows: List<Int>) {
    val payload = JSONObject().apply {
        put("markdown", markdown)
        put("highlightedRows", JSONArray(highlightedRows))
        if (tableName != null) put("tableName", tableName)
    }
    view.evaluateJavascript("window.renderMarkdown($payload);", null)
}
